<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns="http://www.w3.org/TR/REC-html40">

<head>
<meta http-equiv=Content-Type content="text/html; charset=windows-1251">
<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 9">
<meta name=Originator content="Microsoft Word 9">
<link rel=File-List href="./References_files/filelist.xml">
<title>REFERENCES</title>
<!--[if gte mso 9]><xml>
 <o:DocumentProperties>
  <o:Author>Alexander Favorov</o:Author>
  <o:LastAuthor>Alexander Favorov</o:LastAuthor>
  <o:Revision>2</o:Revision>
  <o:TotalTime>3</o:TotalTime>
  <o:Created>2004-11-08T22:49:00Z</o:Created>
  <o:LastSaved>2004-11-08T22:49:00Z</o:LastSaved>
  <o:Pages>1</o:Pages>
  <o:Company>FCCC</o:Company>
  <o:Lines>1</o:Lines>
  <o:Paragraphs>1</o:Paragraphs>
  <o:Version>9.6926</o:Version>
 </o:DocumentProperties>
</xml><![endif]--><!--[if gte mso 9]><xml>
 <w:WordDocument>
  <w:DrawingGridHorizontalSpacing>9.35 pt</w:DrawingGridHorizontalSpacing>
  <w:DrawingGridVerticalSpacing>6.35 pt</w:DrawingGridVerticalSpacing>
  <w:DisplayHorizontalDrawingGridEvery>0</w:DisplayHorizontalDrawingGridEvery>
  <w:DisplayVerticalDrawingGridEvery>2</w:DisplayVerticalDrawingGridEvery>
 </w:WordDocument>
</xml><![endif]-->
<style>
<!--
 /* Font Definitions */
@font-face
	{font-family:Helvetica;
	panose-1:2 11 6 4 2 2 2 2 2 4;
	mso-font-charset:204;
	mso-generic-font-family:swiss;
	mso-font-pitch:variable;
	mso-font-signature:536902279 -2147483648 8 0 511 0;}
 /* Style Definitions */
p.MsoNormal, li.MsoNormal, div.MsoNormal
	{mso-style-parent:"";
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:10.0pt;
	mso-bidi-font-size:12.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	mso-bidi-font-style:italic;}
p.Reference, li.Reference, div.Reference
	{mso-style-name:Reference;
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	text-align:justify;
	mso-pagination:none;
	tab-stops:12.0cm;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.RefHead, li.RefHead, div.RefHead
	{mso-style-name:"Ref Head";
	margin-top:11.3pt;
	margin-right:0cm;
	margin-bottom:3.0pt;
	margin-left:0cm;
	line-height:13.0pt;
	mso-line-height-rule:exactly;
	mso-pagination:widow-orphan;
	mso-outline-level:1;
	font-size:11.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Helvetica;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	text-transform:uppercase;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.RefText, li.RefText, div.RefText
	{mso-style-name:"Ref Text";
	mso-style-parent:"";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:11.35pt;
	margin-bottom:.0001pt;
	text-align:justify;
	text-indent:-11.35pt;
	line-height:11.0pt;
	mso-line-height-rule:exactly;
	mso-pagination:widow-orphan;
	font-size:9.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";}
@page Section1
	{size:612.0pt 792.0pt;
	margin:72.0pt 57.6pt 72.0pt 57.6pt;
	mso-header-margin:36.0pt;
	mso-footer-margin:36.0pt;
	mso-paper-source:0;}
div.Section1
	{page:Section1;}
-->
</style>
</head>

<body lang=EN-US style='tab-interval:36.0pt' bgcolor=white background="back5l24.gif">

<div class=Section1>

<p class=RefText>Bailey,T. and Elkan,C. (1995) Unsupervised learning of
multiple motifs in biopolymers using expectation maximization” <i>Machine
Learning J.</i> <b style='mso-bidi-font-weight:normal'>21</b>, 51-83.</p>

<p class=RefText>Berg,O. G. and von Hippel,P. H. (1987). Selection of DNA
binding sites by regulatory proteins. Statistical-mechanical theory and
application to operators and promoters. <i>J. Mol. Biol.</i> <b
style='mso-bidi-font-weight:normal'>193</b>, 723-750.</p>

<p class=RefText>Besag,J., Green,P., Higdon,D. and Mengersen,K. (1996)<i
style='mso-bidi-font-style:normal'> </i>Bayesian computation and stochastic
systems. <i style='mso-bidi-font-style:normal'>Statistical Science</i>, <b
style='mso-bidi-font-weight:normal'>10 <span style='mso-bidi-font-weight:bold'>(1)</span></b>,
3-66.</p>

<p class=RefText>Bulyk, M. (2003) Computational prediction of
transcription-factor binding site locations. <i>Genome Biology</i>, <b>5</b>:201.</p>

<p class=RefText>Chiang,D.Y., Moses,A.M., Kellis,M., Lander,E.S. and Eisen,M.B.
(2003) Phylogenetically and spatially conserved word pairs associated with
gene-expression changes in yeasts. <i style='mso-bidi-font-style:normal'>Genome
Biol</i>, <b style='mso-bidi-font-weight:normal'>4(R43)</b>.</p>

<p class=RefText>Favorov,A.V., Gelfand,M.S., Mironov,A.A., and Makeev,V.J.
(2002) Yet another digging for DNA motifs Gibbs sampler. <i style='mso-bidi-font-style:
normal'>Proceedings of BGRS 2002</i>, Novisibisk, <b style='mso-bidi-font-weight:
normal'>1</b>, 31-33.</p>

<p class=RefText>Frith,M.C., Li,M.C. and Weng,Z. (2003) Cluster-Buster: Finding
dense clusters of motifs in DNA sequences. <i>Nucleic Acids Res.</i> <b
style='mso-bidi-font-weight:normal'>31(13), </b>3666-8.</p>

<p class=RefText>Gelfand,M.S., Novichkov,P.S., Novichkova,E.S. and Mironov,A.A.
(2000) Comparative analysis of regulatory patterns in bacterial genomes. <i
style='mso-bidi-font-style:normal'>Briefings in Bioinformatics</i>, <b
style='mso-bidi-font-weight:normal'>1</b>, 357–371.</p>

<p class=RefText><span lang=DE style='mso-ansi-language:DE'>Geman,S. and
Geman,D. (1984). </span>Stochastic relaxation, Gibbs distribution and the
Bayesian restoration of images, <i style='mso-bidi-font-style:normal'>IEEE
Transactions on Pattern Analysis and Machine Intelligence</i>. <b
style='mso-bidi-font-weight:normal'>6</b>, 621-641.</p>

<p class=RefText>Gerasimova,A.V., Gelfand,M.S., Makeev,V.J., Mironov,A.A. and
Favorov,A.V. (2004) Primary describtion of ArcA regulon in gamma-proteobacteria
genomes on the base of the regulatory protein binding site computational
recognition. <i>Biophysics</i> [rus.], in press.</p>

<p class=RefText>Harbison,C.T., Gordon,D.B., Lee,T.I., Rinaldi,N.J.,
Macisaac,K.D., Danford,T.W., Hannett,N.M., Tagne,J.B., Reynolds,D.B., Yoo,J.,
Jennings,E.G., Zeitlinger,J., Pokholok,D.K., Kellis,M., Rolfe,P.A.,
Takusagawa,K.T., Lander,E.S., Gifford,D.K., Fraenkel,E. and Young,R.A. (2004)
Transcription regulatory code of a eukaryotic genome. <i style='mso-bidi-font-style:
normal'>Nature</i>, <b style='mso-bidi-font-weight:normal'>431</b>, 99-104.</p>

<p class=RefText>Hertz,G.Z. and Stormo,G.D. (1999) Identifying DNA and protein
patterns with statistically significant alignments of multiple sequences. <i
style='mso-bidi-font-style:normal'>Bioiformatics</i>, <b style='mso-bidi-font-weight:
normal'>15<span style='mso-bidi-font-weight:bold'> (7/8)</span></b>, 563-577</p>

<p class=RefText>Hughes,J.D., Estep,P.W., Tavazoie,S. and Church,G.M. (2000)
Computational identification of cis-regulatory elements associated with groups
of functionally related genes in Saccharomyces cerevisiae. <i style='mso-bidi-font-style:
normal'>J Mol Biol</i>, <b style='mso-bidi-font-weight:normal'>296</b>,
1205-1214.</p>

<p class=RefText>Lawrence,C.E., Altschul,S.F., Boguski,M.S., Liu,J.S.,
Neuwald,A.F., and Wootton,J.C. (1993). Detecting subtle sequence signals: a
Gibbs sampling strategy for multiple alignment. <i style='mso-bidi-font-style:
normal'>Science</i>, <b style='mso-bidi-font-weight:normal'>262</b>, 208-214.</p>

<p class=RefText>Li,H., Rhodius,V., Gross,C. and Siggia,E. (2002) Identification
of the binding sites of regulatory proteins in bacterial genomes. <i>Proc.
Natl. Acad. Sci. </i><b>99,</b> 11772–11777.</p>

<p class=RefText><span lang=DE style='mso-ansi-language:DE'>Liu,J.S. (2001)
Monte Carlo strategies in scientific computing. Springer-Verlag, NY, Berlin,
Heidelberg.<o:p></o:p></span></p>

<p class=RefText>Makeev,V.J., Lifanov,A.P., Nazina,A.G. and Papatsenko,D.A.
(2003) Distance preferences in the arrangement of binding motifs and
hierarchical levels in organization of transcription regulatory information. <i
style='mso-bidi-font-style:normal'>Nucleic Acids Res</i>.<span
style="mso-spacerun: yes">  </span><b style='mso-bidi-font-weight:normal'>31(20),
</b>6016-26.</p>

<p class=RefText>Mironov,A.A., Koonin,E.V., Roytberg,M.A., Gelfand,M.S. (1999)
Computer analysis of transcription regulatory patterns in completely sequenced
bacterial genomes. <i>Nucleic Acids Res.</i> <b>27(14).</b> 2981-2989.</p>

<p class=RefText>Mwangi,M.M., and Siggia,E.D. (2003) Genome wide identification
of regulatory motifs in <i>Bacillus subtilis</i>. <i style='mso-bidi-font-style:
normal'>BMC Bioinformatics</i>. <b style='mso-bidi-font-weight:normal'>4(18)</b>.</p>

<p class=RefText>Pilpel,Y., Sudarsanam,P. and Church, G. (2001) Identifying
regulatory networks by combinatorial analysis of promoter elements. <i>Nat.
Genet.</i>, <b>29</b>, 153-159.</p>

<p class=RefText>Qiu,P., Ding,W., Jiang,Y., Greene,J.R. and Wang, L. (2002)
Computational analysis of composite regulatory elements. <i>Mamm Genome</i>, <b
style='mso-bidi-font-weight:normal'>13(6), </b>327-32.</p>

<p class=RefText>Robert,C.P. (1998) Discretization and MCMC convergence
assessment. Springer-Verlag, NY, Berlin, Heidelberg.</p>

<p class=RefText>Robin,S., Daudin,J.-J., Richard,H., Sagot,M.-F. and Schbath,S.
(2002) Occurrence Probability of Structured Motifs in Random Sequences,<i> J.
Comp. Biol.</i>,<b> 9(6)</b>, 761-773</p>

<p class=RefText>Roth,F.P., Hughes,J.D., Estep,P.W. and Church,G. M. (1998)
Finding DNA regulatory motifs within unaligned noncoding sequences clustered by
whole-genome mRNA quantitation. <i style='mso-bidi-font-style:normal'>Nature
Biotech</i>., <b style='mso-bidi-font-weight:normal'>16</b>, 939-945. </p>

<p class=RefText>Sivia,D.S. (1996) Data Analysis. A Bayesian tutorial.
Clarendon Press, Oxford.</p>

<p class=RefText>Stormo,G.D. and Hartzell,G.W.,3<sup>rd</sup> (1989)
Identifying protein-binding sites from unaligned DNA fragments. <i>Proc Natl
Acad Sci USA</i>, <b>86(4)</b>, 1183-11877.<span style='font-size:10.0pt'> <o:p></o:p></span></p>

<p class=RefText>Thijs,G., Marchall,K., Lescot,M., Rombauts,S., De Moor,B., Rouze,P.
and Moreau, Y. (2002) A Gibbs sampling method to detect over-represented motifs
in the upstream regions of coexpressed genes. <i style='mso-bidi-font-style:
normal'>J. Comp. Biol.,</i> <b style='mso-bidi-font-weight:normal'>9 <span
style='mso-bidi-font-weight:bold'>(2)</span></b>, 447-464.</p>

<p class=RefText>Thompson,W., Rouchka,E.C., Lawrence,C.E. (2003) Gibbs
Recursive Sampler: finding transcription factor binding sites. <i>Nucleic Acids
Res.,</i> <b>31(13)</b>, 3580-3585.</p>

<p class=RefText>Tompa,M., Li,N., Bailey,T.L., Church,G.M., de Moor,B.,
Eskin,E., Favorov,A.V., Frith,M.C., Fu,Y., Kent,W.J., Makeev,V.J.,
Mironov,A.A., Noble,W.S., Pavesi,G., Pesole,G., R&eacute;gnier,M., Simonis,N.,
Sinha,S., Thijs,G., van Helden,J. Vandenbogaert,M., Weng,Z., Workman,C., Ye,C.
and Zhu,Z. (2004) An Assessment of Computational Tools for the Discovery of
Transcription Factor Binding Sites, Nature Biotech., in press.</p>

<p class=RefText>van Helden,J., Rios,A. and Collado-Vides,J. (2000) Discovering
regulatory elements in non-coding sequences by analysis of spaced dyads. <i>Nucleic
Acids Res</i>, <b>28</b>, 1808-1818.</p>

<p class=MsoNormal><span style='mso-ansi-language:EN-US'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p>

</div>

</body>

</html>
